components/eamxx/src/share/io/scream_scorpio_interface.F90

```diff
@@ -1257,6 +1262,121 @@ subroutine set_int_attribute (file_name, attr_name, val)
+      ! be ended when we are done.
+      ierr = PIO_redef(pio_atm_file%pioFileDesc)
+      if (ierr .ne. 0) then
+        call errorHandle("Error while re-opening pio file " // trim(file_name) // ".", -999)
+      endif
+      enddef_needed = .true.
+    endif
+    ierr = PIO_put_att(pio_atm_file%pioFileDesc, PIO_GLOBAL, attr_name, val)
+    if (ierr .ne. 0) then
+      call errorHandle("Error setting global attribute '" // trim(attr_name) &
+                       // "' in pio file " // trim(file_name) // ".", -999)
+    endif
+    if (enddef_needed) then
+      ierr = PIO_enddef(pio_atm_file%pioFileDesc)
+      if (ierr .ne. 0) then
+        call errorHandle("Error while re-closing pio file " // trim(file_name) // ".", -999)
+      endif
+    endif
+  end subroutine set_str_attribute
+!=====================================================================!
+  ! Writes a string attribute to a specific variable in an nc file
+  subroutine set_variable_attribute (file_name, var_name, attr_name, val)
+    use pionfatt_mod, only: PIO_put_att => put_att
+    use pio_nf, only: pio_redef, PIO_inq_att
+
+    character(len=*), intent(in) :: file_name  ! Name of the filename
+    character(len=*), intent(in) :: var_name  ! Name of the filename
+    character(len=*), intent(in) :: attr_name  ! Name of the attribute
+    character(len=*), intent(in) :: val
+    type(pio_atm_file_t), pointer :: pio_atm_file
+    type(hist_var_t), pointer     :: var
+    integer(pio_offset_kind) :: len
+    integer :: ierr,xtype
+    logical :: found, enddef_needed
+
+    call lookup_pio_atm_file(trim(file_name),pio_atm_file,found)
+    if (.not.found) then
+      call errorHandle("PIO Error: can't find pio_atm_file associated with file: "//trim(file_name),-999)
+    endif
+    call get_var(pio_atm_file,var_name,var)
+
+    ! If this attribute does not exist, we need to re-open the nc file for definition,
+    ! then re-close it to put it in data mode again.
+    ! NOTE: this check step is only for pre-NetCDF4 format, where attributes can
+    !       only be defined while in 'define' mode. For NetCDF4/HDF5, attributes
+    !       can be defined at any time.
+    ! TODO: add check on netcdf format, to see if this inq_att shenanigans is needed.
+    ierr = PIO_inq_att(pio_atm_file%pioFileDesc,var%piovar,attr_name,xtype,len)
+    enddef_needed = .false.
+    if (ierr .ne. PIO_NOERR .and. pio_atm_file%is_enddef) then
+      ! In theory, there are several reason why this could fail. However, pio.F90
+      ! does *not* expose all the nc error codes like pio.h does (e.g., no PIO_ENOTATT).
+      ! So we just *assume* that the attribute was not found, and try to define it
+      ! NOTE: We also check if we have ended the definition phase for this file,
+      ! if not we don't need to open it again, and we don't need to flag it to
+      ! be ended when we are done.
+
+      ! TODO - LUCA QUESTION - Is the above statement true?
```

Comment:
@bartgol - tagging you because I wrote this comment inline while adding code.

Reply:
The comment refers to the call to PIO_inq_att right above. When I coded this if statement, the above statement was true (I don't know if it still is, but probably).

Reply:
Uhm, but I guess you are right. We might need to simply check if the file is closed, regardless of whether the att is already defined or not.